flink-core/src/main/java/org/apache/flink/api/common/io/DelimitedInputFormat.java

```diff
@@ -472,6 +498,7 @@ public void open(FileInputSplit split) throws IOException {
 
 		this.offset = splitStart;
 		if (this.splitStart != 0) {
+			setBomFileCharset(split);
```

Comment:
I have two questions about this commit, as follows:
For the first suggestion, I feel that users often cannot know the encoding of the file accurately. For example: file encoding `UTF-16LE`, with bom header, user-specified encoding `UTF-16BE` will report an error. And there is bom UTF encoding I believe will be the majority. So I think it is necessary to do the bom code detection first, which is better for the user experience.
For the fourth recommendation, the seek of `GenericCsvInputFormat` cannot be seek to position 0. It calls the `seek` method of `InputStreamFSInputWrapper`. This method cannot currently seek to position 0.

Reply:
I am suggesting that we only do a BOM check if the user did not configure a charset or if the user explicitly configured an UTF charset. In case the user explicitly configured a different charset, we do not check the BOM because it could also be valid data.

`InputStreamFSInputWrapper`'s limitation in seeking makes it a bit tricky to read the first bytes of a file, since the stream is opened and seeked in `FileInputFormat.open()`. Maybe we can implement a protected method `FileInputFormat.readFileHeader()` that is called after the stream was opened. `DelimitedInputFormat` can then override the method to check for the BOM. By default, the method should not do anything.

Reply:
`DelimitedInputFormat` is filled with a `readBuffer` variable when calling `fillBuffer(0)`. We can copy 4 bytes from readBuffer. It should not have a big impact. As follows: 
`System.arraycopy(this.readBuffer, 0, bomBuffer, 0, 4);`

Reply:
@fhueske  Adding `FileInputFormat.readFileHeader()` to `FileInputFormat` still needs to get the 4 bytes of the bom header through the stream. I think it's okay to open the `stream` in `DelimitedInputFormat` and then process it. Also for the Stream of `InputStreamFSInputWrapper`'s I need to open and read 4 bytes and then close the stream. `But fillBuffer(0)` will also do the open and close operations of the stream. This is my question.